File: ios/ExpoDouyinOpenSDKModule.swift
import ExpoModulesCore
import DouyinOpenSDK
import os

public class ExpoDouyinOpenSDKModule: Module {
    
    let logger = Logger()
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoDouyinOpenSDK')` in JavaScript.
    Name("ExpoDouyinOpenSDK")

      Function("initSDK") { (appId: String) -> Bool in
          logger.info("Start to init Douyin OpenSDK with client key: \(appId)")
          return DouyinOpenSDKApplicationDelegate.sharedInstance().registerAppId(appId)
      }
      
      AsyncFunction("authorize") { (acceptWhiteList: Bool, promise: Promise) in
          let appId = DouyinOpenSDKApplicationDelegate.sharedInstance().appId()
          if appId == nil {
              logger.error("SDK not initialized!!")
              promise.reject(Exception(name: "unintialized", description: "SDK not initialized!!"))
          }
          let request: DouyinOpenSDKAuthRequest = DouyinOpenSDKAuthRequest()
          request.permissions = if (acceptWhiteList) {
              ["user_info", "trial.whitelist"]
          } else {
              ["user_info"]
          }
          let currentVC: UIViewController? = Utilities.getCurrentViewController()
          if currentVC != nil {
              request.send(currentVC!) { resp in
                  switch resp?.errCode {
                  case .success, .success20000:
                      let grantedPermissionsString = resp!.grantedPermissions?.array
                              .compactMap { $0 as? String }
                              .joined(separator: ",")
                      let authResult = AuthResult(
                        authType: AuthType.SUCCESS.toString(),
                        authCode: resp!.code ?? "",
                        grantedPermissions: grantedPermissionsString ?? "",
                        errorCode: 0,
                        errorMsg: resp?.errString ?? ""
                      )
                      promise.resolve(authResult.toDictionary())
                      
                  case .errorCodeUserCanceled:
                      let grantedPermissionsString = resp?.grantedPermissions?.array
                          .compactMap { $0 as? String }
                          .joined(separator: ",")
                      let authResult = AuthResult(
                        authType: AuthType.USER_CANCELED.toString(),
                          authCode: resp?.code ?? "",
                          grantedPermissions: grantedPermissionsString ?? "",
                          errorCode: -2,
                          errorMsg: resp?.errString ?? ""
                      )
                      promise.resolve(authResult.toDictionary())

                  default:
                      let grantedPermissionsString = resp?.grantedPermissions?.array
                          .compactMap { $0 as? String }
                          .joined(separator: ",")
                      let authResult = AuthResult(
                        authType: AuthType.FAILED.toString(),
                          authCode: resp?.code ?? "",
                          grantedPermissions: grantedPermissionsString ?? "",
                          errorCode: Utilities.getErrorCodeValue(errorCode: resp!.errCode.rawValue).intValue,
                          errorMsg: resp?.errString ?? ""
                      )
                      promise.resolve(authResult.toDictionary())
                  }
              }
          } else {
              promise.reject(Exception(name: "no_view_controller", description: "Could not find current UIViewController"))
          }
          
      }
      
      Function("isDouyinInstalled") { () -> Bool in
          return DouyinOpenSDKApplicationDelegate.sharedInstance().isAppInstalled()
      }

  }
}

class Utilities {
    // 获取当前显示的 ViewController
    static func getCurrentViewController() -> UIViewController? {
        guard let rootViewController = UIApplication.shared.keyWindow?.rootViewController else {
            return nil
        }
        return getTopViewController(from: rootViewController)
    }

    // 获取顶部 ViewController
    static func getTopViewController(from rootViewController: UIViewController) -> UIViewController {
        if let presentedViewController = rootViewController.presentedViewController {
            return getTopViewController(from: presentedViewController)
        } else if let navigationController = rootViewController as? UINavigationController {
            return getTopViewController(from: navigationController.topViewController ?? navigationController)
        } else if let tabBarController = rootViewController as? UITabBarController {
            return getTopViewController(from: tabBarController.selectedViewController ?? tabBarController)
        } else {
            return rootViewController
        }
    }

    // 将错误码转换为 NSNumber
    static func getErrorCodeValue(errorCode: Int) -> NSNumber {
        return NSNumber(value: errorCode)
    }
}


enum AuthType {
    /**
     * 授权成功
     */
    case SUCCESS
    /**
     * 用户取消授权
     */
    case USER_CANCELED
    /**
     * 授权失败
     */
    case FAILED
    
    public func toString() -> String {
        switch self {
        case .SUCCESS:
            return "SUCCESS"
        case .USER_CANCELED:
            return "USER_CANCELED"
        case .FAILED:
            return "FAILED"
        }
    }
}

class AuthResult {
    var authType: String = ""
    var authCode: String = ""
    var grantedPermissions: String = ""
    var errorCode: Int = 0
    var errorMsg: String = ""
    
    init(authType: String, authCode: String, grantedPermissions: String, errorCode: Int, errorMsg: String) {
        self.authType = authType
        self.authCode = authCode
        self.grantedPermissions = grantedPermissions
        self.errorCode = errorCode
        self.errorMsg = errorMsg
    }
    
    public func toDictionary() -> NSMutableDictionary {
        let dict = NSMutableDictionary()
        dict["authType"] = self.authType
        dict["authCode"] = self.authCode
        dict["grantedPermissions"] = self.grantedPermissions
        dict["errorCode"] = self.errorCode
        dict["errorMsg"] = self.errorMsg
        return dict
    }
}

File: android/src/main/java/com/lnyynet/expomodule/dysdk/ExpoDouyinOpenSDKModule.kt
package com.lnyynet.expomodule.dysdk

import android.util.Log
import com.bytedance.sdk.open.aweme.authorize.model.Authorization
import com.bytedance.sdk.open.douyin.DouYinOpenApiFactory
import com.bytedance.sdk.open.douyin.DouYinOpenConfig
import com.lnyynet.expomodule.dysdk.common.AuthResult
import com.lnyynet.expomodule.dysdk.common.toWritableMap
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoDouyinOpenSDKModule : Module() {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoDouyinOpenSDK')` in JavaScript.
    Name("ExpoDouyinOpenSDK")

    // Defines a JavaScript synchronous function that runs the native code on the JavaScript thread.
    Function("initSDK") { clientKey: String ->
      return@Function DouYinOpenApiFactory.init(DouYinOpenConfig(clientKey)).also {
        isInitialized = it
      }
    }

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    AsyncFunction("authorize") { acceptWhiteList: Boolean, promise: Promise ->
      // 判断是否已经初始化 SDK
        if (!isInitialized) {
            promise.reject("SDK_NOT_INITIALIZED", "SDK is not initialized, please call initSDK first", null)
            return@AsyncFunction
        }
      authorizePromise = promise
      // 调用 SDK 发起授权请求，这里会跳转到外部应用
      val douyinOpenApi = DouYinOpenApiFactory.create(appContext.currentActivity)

      val request = Authorization.Request()
      request.scope = if (acceptWhiteList) "user_info,trial.whitelist" else "user_info" // 用户授权时必选权限
      request.callerLocalEntry = "com.lnyynet.expomodule.dysdk.douyinapi.DouYinEntryActivity" // 第三方指定自定义的回调类 Activity
      // 优先使用抖音app进行授权，如果抖音app因版本或者其他原因无法授权，则使用web页授权
      request.state = "ww" // 用于保持请求和回调的状态，授权请求后原样带回给第三方
      Log.d("ExpoDouyinOpenSDK", "pending authorize: $request")
      douyinOpenApi.authorize(request)
    }

    Function("isDouyinInstalled") {
      val douyinOpenApi = DouYinOpenApiFactory.create(appContext.currentActivity)
      return@Function douyinOpenApi.isAppInstalled
    }

  }

  companion object {
    private var isInitialized = false // 用于保存 SDK 初始化状态

    private var authorizePromise: Promise? = null // 用于保存 Promise 对象

    // 当授权结果返回时调用该方法
    fun onAuthResult(result: AuthResult) {
      Log.d("ExpoDouyinOpenSDK", "onAuthResult: $result")
      if (authorizePromise != null) {
        authorizePromise!!.resolve(result.toWritableMap()) // 返回授权结果
        authorizePromise = null // 清空 Promise 防止内存泄漏
      } else {
        // 如果没有保存 Promise 对象，说明授权请求没有发起，这里可以忽略
        // 打印一个警告日志，提示回调被意外触发
        Log.w("ExpoDouyinOpenSDK", "authorize promise is null, pls check if you have called authorize method")
      }
    }
  }
}
